src/codegen/templates/_grpc_interpreter.py.mako

```diff
@@ -221,10 +230,24 @@ def _assign_numpy_array(numpy_array, grpc_array):
     grpc_array_size = len(grpc_array)
     if isinstance(grpc_array, bytes):
         assert numpy_array.nbytes >= grpc_array_size
-        numpy_array.flat[:grpc_array_size] = numpy.frombuffer(grpc_array, dtype=numpy_array.dtype)
+        if numpy_array.ndim > 1:
+            (number_of_channels, number_of_samples_per_channel) = numpy_array.shape
+            read_array = numpy.frombuffer(grpc_array, dtype=numpy_array.dtype).reshape((number_of_channels, samps_per_chan_read))
+            numpy_array[:, :samps_per_chan_read] = read_array
+        elif num_samps_per_chan==1 and numpy_array.size<=samps_per_chan_read:
+            numpy_array.flat[0:numpy_array.size] = numpy.frombuffer(grpc_array, dtype=numpy_array.dtype)
+        else:
+            numpy_array.flat[0:samps_per_chan_read] = numpy.frombuffer(grpc_array, dtype=numpy_array.dtype)
     else:
         assert numpy_array.size >= grpc_array_size
-        numpy_array.flat[:grpc_array_size] = grpc_array
+        if numpy_array.ndim > 1:
+            (number_of_channels, number_of_samples_per_channel) = numpy_array.shape
+            read_array = numpy.array(grpc_array).reshape((number_of_channels, samps_per_chan_read))
+            numpy_array[:, :samps_per_chan_read].flat[:grpc_array_size] = grpc_array
+        elif num_samps_per_chan==1 and numpy_array.size<=samps_per_chan_read:
+            numpy_array.flat[0:numpy_array.size] = grpc_array[0:numpy_array.size]
```

Comment:
What is this middle case intended to do?

It looks like it's an attempt to handle N-channel 1-sample, but it won't work because samps_per_chan_read will be 1 in that case, not N.

It won't handle flat arrays (raw reads) either because those typically have num_samps_per_chan > 1.

Reply:
It was supposed to handle N-channel 1-sample case but as you pointed out it needs rework. This will be addressed in [Bug 2420742](https://dev.azure.com/ni/DevCentral/_workitems/edit/2420742): gRPC interpreter doesn't retain the initial array values when converting flattened array into numpy array.